jax/experimental/jax2tf/jax2tf.py

```diff
@@ -334,135 +334,255 @@ def convert(fun_jax: Callable,
     finally:
-      _thread_local_state.shape_env = ()
-      _thread_local_state.enable_xla = prev_enable_xla
-      _thread_local_state.include_xla_op_metadata = prev_include_xla_op_metadata
-      _thread_local_state.tf_outer_name_scope = prev_tf_outer_name_scope
+      impl.restore_context()
 
-    out_flat_tf = [tf.identity(x, "jax2tf_out") for x in out_flat_tf]
-    out_tf = tree_util.tree_unflatten(out_tree_thunk(), out_flat_tf)
+    outs_flat_tf = [tf.identity(x, "jax2tf_out") for x in outs_flat_tf]
+    out_tf = tree_util.tree_unflatten(outs_tree, outs_flat_tf)
     return out_tf
 
   return converted_fun_tf
 
+class SerializationImpl:
+  """Implementation details for jax2tf serialization.
+
+  Abstract superclass for subclassing.
+  """
+  def setup_context(self):
+    """Called in the resulting TF function, before any other method."""
+    raise NotImplementedError
+
+  def restore_context(self):
+    """Called in the resulting TF function, after conversion is done."""
+    raise NotImplementedError
+
+  def run_fun_tf(self,
+                 args_flat_tf: Sequence[TfVal]
+                 ) -> Tuple[Sequence[TfVal], Sequence[core.ShapedArray], tree_util.PyTreeDef]:
+    """Runs the resulting TF function.
+
+    Args:
+      args_flat_tf: a flat tuple of tf.Tensor arguments
+
+    Returns: a tuple with:
+      outs_tfs: a flat tuple of tf.Tensor results
+      outs_avals: a flat tuple of JAX abstract values for the underlying JAX
+        function.
+      outs_tree: the PyTreeDef for the outputs
+    """
+    raise NotImplementedError
+
+  def run_vjp_fun_tf(self,
+                     vjp_args_flat_tf: Sequence[TfVal],
+                     outs_avals: Sequence[core.AbstractValue]) -> Sequence[TfVal]:
+    """Runs the VJP function as a TF function.
+
+    Args:
+      vjp_args_flat_tf: the flattened sequence of tf.Tensor, including the
+          primal arguments followed by the output cotangents.
+      outs_avals: the flattened primal outputs avals
+
+    Returns: the flattened sequence of input cotangents.
+    """
+    raise NotImplementedError
+
+
+class NativeSerializationImpl(SerializationImpl):
+  def __init__(self, fun_jax, *,
+               args_specs, kwargs_specs,
+               native_serialization_platforms: Sequence[str],
+               native_serialization_strict_checks: bool):
+    self.fun_jax = fun_jax
+    self.args_specs = args_specs
+    self.kwargs_specs = kwargs_specs
+    self.native_serialization_strict_checks = native_serialization_strict_checks
+    if native_serialization_platforms:
+      self.lowering_platform: Optional[str] = native_serialization_platforms[0]
+    else:
+      self.lowering_platform = None
+
+  def setup_context(self):
+    self.exported = jax_export.export(
+        self.fun_jax,
+        lowering_platform=self.lowering_platform,
+        strict_checks=self.native_serialization_strict_checks
+    )(*self.args_specs, **self.kwargs_specs)
+
+  def restore_context(self):
+    pass
```

Comment:
Not fully understand setup_context and restore_context. My understanding is that they are reverse operations. Should we need add `del self.exported` here since we add it in setup_context ?

Reply:
I have changed the name to `before_conversion` and `after_conversion`. The goal is to be able to setup/restore global state that may be needed. We do not need to restore changes to `self`